sklearn/tree/_splitter.pyx

```diff
@@ -710,39 +820,71 @@ cdef class DensePartitioner:
         cdef DTYPE_t[::1] feature_values
         cdef SIZE_t start
         cdef SIZE_t end
+        cdef SIZE_t n_missing
+        cdef const unsigned char[::1] missing_mask
 
     def __init__(
         self,
         const DTYPE_t[:, :] X,
         SIZE_t[::1] samples,
         DTYPE_t[::1] feature_values,
+        const unsigned char[::1] missing_mask,
     ):
         self.X = X
         self.samples = samples
         self.feature_values = feature_values
+        self.missing_mask = missing_mask
 
     cdef inline void init_node_split(self, SIZE_t start, SIZE_t end) noexcept nogil:
         """Initialize splitter at the beginning of node_split."""
         self.start = start
         self.end = end
+        self.n_missing = 0
 
     cdef inline void sort_samples_and_feature_values(
         self, SIZE_t current_feature
     ) noexcept nogil:
-        """Simultaneously sort based on the feature_values."""
+        """Simultaneously sort based on the feature_values.
+
+        Returns the number of missing values in feature_values.
+        """
         cdef:
-            SIZE_t i
+            SIZE_t i, current_end
             DTYPE_t[::1] feature_values = self.feature_values
             const DTYPE_t[:, :] X = self.X
             SIZE_t[::1] samples = self.samples
+            SIZE_t n_missing = 0
+            const unsigned char[::1] missing_mask = self.missing_mask
 
         # Sort samples along that feature; by
         # copying the values into an array and
         # sorting the array in a manner which utilizes the cache more
         # effectively.
-        for i in range(self.start, self.end):
-            feature_values[i] = X[samples[i], current_feature]
-        sort(&feature_values[self.start], &samples[self.start], self.end - self.start)
+        if missing_mask is not None and missing_mask[current_feature]:
+            i, current_end = self.start, self.end - 1
+            # Missing values are placed at the end and do not participate in the sorting.
+            while i <= current_end:
+                # Finds the lowest current_end that is not missing
```

Comment:
```suggestion
                # Finds the right-most value that is not missing so that
                # it can be swapped with missing values at its left.
```